cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -441,33 +455,26 @@ private boolean isConcurrentRunsAllowed(ProgramType type) {
     return EnumSet.of(ProgramType.WORKFLOW, ProgramType.MAPREDUCE).contains(type);
   }
 
-  @Nullable
-  private ProgramRuntimeService.RuntimeInfo findRuntimeInfo(ProgramId programId,
+  private List<ProgramRuntimeService.RuntimeInfo> findRuntimeInfo(ProgramId programId,
                                                             @Nullable String runId) throws BadRequestException {
-    Map<RunId, ProgramRuntimeService.RuntimeInfo> runtimeInfos = runtimeService.list(programId.getType());
-
     if (runId != null) {
       RunId run;
       try {
         run = RunIds.fromString(runId);
       } catch (IllegalArgumentException e) {
         throw new BadRequestException("Error parsing run-id.", e);
       }
-      return runtimeInfos.get(run);
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = runtimeService.lookup(programId.toId(), run);
+      return runtimeInfo == null ? new ArrayList<ProgramRuntimeService.RuntimeInfo>() :
+        Collections.singletonList(runtimeInfo);
     }
-
-    return findRuntimeInfo(programId);
+    return new ArrayList<>(runtimeService.list(programId.toId()).values());
```

Comment:
Because I need to return a List(), is there a better way to convert collection to list?